gemini/use-cases/retail/product_recognition_from_image.ipynb

```diff
@@ -0,0 +1,506 @@
+        "    import IPython\n",
+        "\n",
+        "    app = IPython.Application.instance()\n",
+        "    app.kernel.do_shutdown(True)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "id": "ot7iCovUlEK6",
+      "metadata": {
+        "id": "ot7iCovUlEK6"
+      },
+      "outputs": [],
+      "source": [
+        "# @title Authetication\n",
+        "import sys\n",
+        "\n",
+        "if \"google.colab\" in sys.modules:\n",
+        "    from google.colab import auth\n",
+        "\n",
+        "    auth.authenticate_user()"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "id": "fL3GwbKKlEtE",
+      "metadata": {
+        "id": "fL3GwbKKlEtE"
+      },
+      "outputs": [],
+      "source": [
+        "# @title Setup Vertex project.\n",
+        "PROJECT_ID = \"[your-project-id]\"  # @param {type:\"string\"}\n",
+        "LOCATION = \"us-central1\"  # @param {type:\"string\"}\n",
+        "\n",
+        "import vertexai\n",
+        "\n",
+        "vertexai.init(project=PROJECT_ID, location=LOCATION)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "id": "fHVclx2ElUEF",
+      "metadata": {
+        "id": "fHVclx2ElUEF"
+      },
+      "source": [
+        "# Product Recognition using Gemini Demo"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 9,
+      "id": "xxlXl4hiDKPeITgNTUBMGAFP",
+      "metadata": {
+        "cellView": "form",
+        "id": "xxlXl4hiDKPeITgNTUBMGAFP"
+      },
+      "outputs": [],
+      "source": [
+        "from io import BytesIO\n",
+        "\n",
+        "from IPython.display import display\n",
+        "from PIL import Image\n",
+        "from googleapiclient.discovery import build  # For google custom search engine.\n",
+        "\n",
+        "# @title [Library] Gemini based Product Image Search/Recognition.\n",
+        "import ipywidgets as widgets\n",
+        "import requests\n",
+        "from vertexai.preview.generative_models import GenerationConfig, GenerativeModel, Part\n",
+        "\n",
+        "# Google Custom search engine developer key.\n",
+        "# To get the key, follow the instructions at\n",
+        "# https://developers.google.com/custom-search/v1/overview\n",
+        "YOUR_DEVELOPER_KEY = \"\"  # @param {\"type\": \"string\"}\n",
+        "\n",
+        "\n",
+        "def get_image_mime_from_uri(image_uri: str) -> str:\n",
+        "    \"\"\"Get the mime type from the image uri.\"\"\"\n",
+        "    if image_uri.endswith(\".png\"):\n",
+        "        return \"image/png\"\n",
+        "    elif image_uri.endswith(\".gif\"):\n",
+        "        return \"image/gif\"\n",
+        "    elif image_uri.endswith(\".svg\"):\n",
+        "        return None  # not a normal image.\n",
+        "    else:\n",
+        "        # Assume JPEG as the default mime\n",
+        "        return \"image/jpeg\"\n",
```

Comment:
There are libraries that can do this. https://docs.python.org/3/library/mimetypes.html